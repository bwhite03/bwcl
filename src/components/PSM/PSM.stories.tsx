import React from 'react';
import { Story, Meta } from '@storybook/react';
import PSM, { PSMProps } from './PSM';

export default {
  title: 'PSM',
  component: PSM,
} as Meta;

const Template: Story<PSMProps> = (args) => <PSM {...args} />;
export const PSMExample = Template.bind({});
PSMExample.args = { userid: 1 };
