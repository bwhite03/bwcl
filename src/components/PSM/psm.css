.password-title-container {
  width: 100%;
  display: grid;
  grid-template-columns: 1fr 1fr;
}

.password-title {
  text-align: left;
  font-weight: bold;
}

.password-warning {
  color: red;
  text-align: right;
}

.password-instructions {
  text-align: center;
  font-size: 0.8em;
}

.password-bar-container-hidden {
  display: none;
}

.password-bar-container {
  height: 10px;
  display: grid;
  grid-template-columns: 1fr 1fr 1fr 1fr;
  margin-bottom: 10px;
  margin-top: 10px;
}

.password-bar {
  border: 1px solid black;
  border-radius: 5px;
  margin-left: 2px;
  margin-right: 2px;
}

.empty-password-bar {
  background-color: #eee;
}

.bad-password {
  background-color: darkred;
}

.ok-password {
  background-color: orangered;
}

.decent-password {
  background-color: orange;
}

.better-password {
  background-color: yellowgreen;
}

.good-password {
  background-color: green;
}

.password-input-1 {
  width: 90%;
  margin: 8px auto;
  display: block;
  border-radius: 5px;
  outline: none;
}

.password-input-1:focus {
  outline: none;
}

.btn-outline-dark {
  padding: 5px 8px;
  border-radius: 10px;
  margin-top: 5px;
  border: 2px solid black;
  cursor: pointer;
  transition: all;
}

.btn-outline-dark:hover {
  background-color: rgb(167, 161, 161);
  color: #fff;
}
