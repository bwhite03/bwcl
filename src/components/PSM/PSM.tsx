import React, { useState, useEffect } from 'react';
import './psm.css';

export interface PSMProps {
  userid: number;
  confirmPasswordChange: (password: string) => void;
}

const PSM = (props: PSMProps) => {
  const [password, setPassword] = useState('');
  const [confirmPassword, setConfirmPassword] = useState('');
  const [passwordLength, setPasswordLength] = useState(0);
  const [buttonActive, setButtonActive] = useState(false);
  const [passwordWarning, setPasswordWarning] = useState(
    'Password is required'
  );
  const [barOneClass, setBarOneClass] = useState('');
  const [barTwoClass, setBarTwoClass] = useState('');
  const [barThreeClass, setBarThreeClass] = useState('');
  const [barFourClass, setBarFourClass] = useState('');
  const [masterScore, setMasterScore] = useState(0);

  const passwordMinimum = 7;
  let manatory = 0;

  useEffect(() => {
    calculateScore();
    if (passwordLength < passwordMinimum && password.length > 0) {
      setPasswordWarning('Password is too short');
    } else if (passwordLength > passwordMinimum) {
      setPasswordWarning('');
    }
  }, [password]);

  const calculateScore = () => {
    let intScore = 0;
    manatory = 0;
    intScore = passwordLength;

    if (passwordLength > 0 && passwordLength <= 4) {
      intScore += passwordLength;
    } else if (passwordLength >= 5 && passwordLength <= 7) {
      intScore += 6;
    } else if (passwordLength >= 8 && passwordLength <= 15) {
      intScore += 12;
    } else if (passwordLength >= 16) {
      intScore += 18;
    }

    if (password.match(/[a-z]/)) {
      intScore += 5;
      manatory++;
    }

    if (password.match(/[A-Z]/)) {
      intScore += 5;
      manatory++;
    }

    if (password.match(/\d/)) {
      intScore += 5;
      manatory++;
    }

    if (password.match(/[!,@,#,$,^,&,*,?,_,~]/)) {
      intScore += 5;
      manatory++;
    }

    if (password.match(/(?:.*?[!,@,#,$,^,&,*,_,~]){2}/)) {
      intScore += 5;
    }

    if (password.match(/(?=.*[a-z])(?=.*[A-Z])/)) {
      intScore += 5;
    }

    // if (
    //   password.match(/(?=.*\d)(?=.*[a-z])(?=.*[A-Z])(?=.*[!,@,#,$,^,&,*,_,~])/)
    // ) {
    //   intScore += 5;
    // }

    switch (manatory) {
      case 0:
        setBarOneClass('bad-password');
        setBarTwoClass('');
        setBarThreeClass('');
        setBarFourClass('');
        break;
      case 1:
        setBarOneClass('bad-password');
        setBarTwoClass('');
        setBarThreeClass('');
        setBarFourClass('');
        break;
      case 2:
        setBarOneClass('ok-password');
        setBarTwoClass('bad-password');
        setBarThreeClass('');
        setBarFourClass('');
        break;
      case 3:
        setBarOneClass('decent-password');
        setBarTwoClass('ok-password');
        setBarThreeClass('bad-password');
        setBarFourClass('');
        break;
      case 4:
        setBarOneClass('better-password');
        setBarTwoClass('decent-password');
        setBarThreeClass('ok-password');
        setBarFourClass('bad-password');
        break;
    }

    setMasterScore(intScore * 2);

    if (passwordLength > passwordMinimum && manatory >= 4) {
      setBarOneClass('good-password');
      setBarTwoClass('good-password');
      setBarThreeClass('good-password');
      setBarFourClass('good-password');
      setButtonActive(true);
    } else {
      setButtonActive(false);
    }
  };

  const handlePasswordChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setPassword(e.target.value);
    setPasswordLength(e.target.value.length);
  };

  const handleConfirmChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setConfirmPassword(e.target.value);
  };

  const handleClick = (e: React.MouseEvent<HTMLButtonElement>) => {
    e.preventDefault();
    if (password !== confirmPassword) {
      setPasswordWarning('Passwords to not match');
      return;
    }

    if (!props.userid || props.userid === 0) {
      setPasswordWarning('Please select a user');
      return;
    }
    props.confirmPasswordChange(password);
  };

  return (
    <div className="password-wrapper" style={{ textAlign: 'center' }}>
      <div className="password-title-container">
        <div className="password-title">Password</div>
        <div className="password-warning" id="password-warning">
          {passwordWarning}
        </div>
      </div>
      <div className="password-instructions">
        <div>
          You need at least one special character, one number, one uppercase
          letter and one lowercase character.
        </div>
        <div>
          {' '}
          {` Also a minium of ${passwordMinimum}
        characters.`}
        </div>
      </div>

      <div
        className={
          passwordLength === 0
            ? 'password-bar-container-hidden'
            : 'password-bar-container'
        }
      >
        <div className={`password-bar ${barOneClass}`} id="password-one"></div>
        <div className={`password-bar ${barTwoClass}`} id="password-one"></div>
        <div
          className={`password-bar ${barThreeClass}`}
          id="password-one"
        ></div>
        <div className={`password-bar ${barFourClass}`} id="password-one"></div>
      </div>
      <input
        className="password-input-1"
        type="password"
        autoComplete="off"
        onChange={handlePasswordChange}
        value={password}
        placeholder="Enter Password"
        name="password"
      />
      <input
        className="password-input-1"
        type="password"
        autoComplete="off"
        onChange={handleConfirmChange}
        value={confirmPassword}
        placeholder="Confirm Password"
        name="confirm"
      />
      <div className="int-score">{`Score ${masterScore}`}</div>
      {!buttonActive ? (
        <div></div>
      ) : (
        <button className="btn btn-outline-dark" onClick={handleClick}>
          Submit
        </button>
      )}
    </div>
  );
};

export default PSM;
