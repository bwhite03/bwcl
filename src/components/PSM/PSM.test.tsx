import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import PSM, { PSMProps } from './PSM';

describe('PSM', () => {
  test('should render correctly', () => {
    const testFn = jest.fn();
    const { container } = render(
      <PSM userid={1} confirmPasswordChange={testFn} />
    );

    expect(container).toMatchSnapshot();
  });

  test('should handle a single character', () => {
    const testFn = jest.fn();
    const { getByPlaceholderText } = render(
      <PSM userid={1} confirmPasswordChange={testFn} />
    );
    userEvent.type(screen.getByPlaceholderText('Enter Password'), '!');
    const input = screen.getByPlaceholderText(
      'Enter Password'
    ) as HTMLInputElement;
    expect(input.value).toEqual('!');
  });

  test('should handle a three character', () => {
    const testFn = jest.fn();
    const { getByPlaceholderText } = render(
      <PSM userid={1} confirmPasswordChange={testFn} />
    );
    userEvent.type(screen.getByPlaceholderText('Enter Password'), '!Ss');
    const input = screen.getByPlaceholderText(
      'Enter Password'
    ) as HTMLInputElement;
    expect(input.value).toEqual('!Ss');
  });

  test('should handle a five character', () => {
    const testFn = jest.fn();
    const { getByPlaceholderText } = render(
      <PSM userid={1} confirmPasswordChange={testFn} />
    );
    userEvent.type(screen.getByPlaceholderText('Enter Password'), '!Ssss');
    const input = screen.getByPlaceholderText(
      'Enter Password'
    ) as HTMLInputElement;
    expect(input.value).toEqual('!Ssss');
  });

  test('should handle a seven character', () => {
    const testFn = jest.fn();
    const { getByPlaceholderText } = render(
      <PSM userid={1} confirmPasswordChange={testFn} />
    );
    userEvent.type(screen.getByPlaceholderText('Enter Password'), '!Ssss11');
    const input = screen.getByPlaceholderText(
      'Enter Password'
    ) as HTMLInputElement;
    expect(input.value).toEqual('!Ssss11');
  });

  test('should handle a password greater than 17 characters', () => {
    const testFn = jest.fn();
    const { getByPlaceholderText } = render(
      <PSM userid={1} confirmPasswordChange={testFn} />
    );
    userEvent.type(
      screen.getByPlaceholderText('Enter Password'),
      '!Ssss1111111111111'
    );
    const input = screen.getByPlaceholderText(
      'Enter Password'
    ) as HTMLInputElement;
    expect(input.value).toEqual('!Ssss1111111111111');
  });

  test('should handle a two adjecent special characters', () => {
    const testFn = jest.fn();
    const { getByPlaceholderText } = render(
      <PSM userid={1} confirmPasswordChange={testFn} />
    );
    userEvent.type(
      screen.getByPlaceholderText('Enter Password'),
      '!!@5Teams5!!@'
    );
    const input = screen.getByPlaceholderText(
      'Enter Password'
    ) as HTMLInputElement;
    expect(input.value).toEqual('!!@5Teams5!!@');
  });

  test('should handle submit button click', () => {
    const testFn = jest.fn();
    const { container, getByPlaceholderText } = render(
      <PSM userid={1} confirmPasswordChange={testFn} />
    );
    userEvent.type(
      screen.getByPlaceholderText('Enter Password'),
      '!!@5Teams5!!@'
    );
    const input = screen.getByPlaceholderText(
      'Enter Password'
    ) as HTMLInputElement;
    const btn = container.querySelector('button');

    expect(btn).toBeVisible();

    if (btn) {
      fireEvent.click(btn);
    }
  });

  test('should handle no user for props', () => {
    const testFn = jest.fn();
    const { container, getByPlaceholderText } = render(
      // @ts-ignore
      <PSM confirmPasswordChange={testFn} />
    );
    userEvent.type(
      screen.getByPlaceholderText('Enter Password'),
      '!!@5Teams5!!@'
    );
    userEvent.type(
      screen.getByPlaceholderText('Confirm Password'),
      '!!@5Teams5!!@'
    );
    const btn = container.querySelector('button');
    if (btn) {
      fireEvent.click(btn);
      const text = screen.getByText('Please select a user');
      expect(text).toBeVisible();
    }
  });

  test('should handle userid of 0', () => {
    const testFn = jest.fn();
    const { container, getByPlaceholderText } = render(
      // @ts-ignore
      <PSM confirmPasswordChange={testFn} userid={0} />
    );
    userEvent.type(
      screen.getByPlaceholderText('Enter Password'),
      '!!@5Teams5!!@'
    );
    userEvent.type(
      screen.getByPlaceholderText('Confirm Password'),
      '!!@5Teams5!!@'
    );
    const btn = container.querySelector('button');
    if (btn) {
      fireEvent.click(btn);
      const text = screen.getByText('Please select a user');
      expect(text).toBeVisible();
    }
  });

  test('should successfully submit new password', () => {
    const testFn = jest.fn();
    const { container, getByPlaceholderText } = render(
      // @ts-ignore
      <PSM confirmPasswordChange={testFn} userid={1} />
    );
    userEvent.type(
      screen.getByPlaceholderText('Enter Password'),
      '!!@5Teams5!!@'
    );
    userEvent.type(
      screen.getByPlaceholderText('Confirm Password'),
      '!!@5Teams5!!@'
    );
    const btn = container.querySelector('button');
    if (btn) {
      fireEvent.click(btn);
    }
  });
});
