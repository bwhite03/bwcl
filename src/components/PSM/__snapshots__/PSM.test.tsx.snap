// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PSM should render correctly 1`] = `
<div>
  <div
    class="password-wrapper"
    style="text-align: center;"
  >
    <div
      class="password-title-container"
    >
      <div
        class="password-title"
      >
        Password
      </div>
      <div
        class="password-warning"
        id="password-warning"
      >
        Password is required
      </div>
    </div>
    <div
      class="password-instructions"
    >
      <div>
        You need at least one special character, one number, one uppercase letter and one lowercase character.
      </div>
      <div>
         
         Also a minium of 7
        characters.
      </div>
    </div>
    <div
      class="password-bar-container-hidden"
    >
      <div
        class="password-bar bad-password"
        id="password-one"
      />
      <div
        class="password-bar "
        id="password-one"
      />
      <div
        class="password-bar "
        id="password-one"
      />
      <div
        class="password-bar "
        id="password-one"
      />
    </div>
    <input
      autocomplete="off"
      class="password-input-1"
      name="password"
      placeholder="Enter Password"
      type="password"
      value=""
    />
    <input
      autocomplete="off"
      class="password-input-1"
      name="confirm"
      placeholder="Confirm Password"
      type="password"
      value=""
    />
    <div
      class="int-score"
    >
      Score 0
    </div>
    <div />
  </div>
</div>
`;
